src/front/wgsl/mod.rs

```diff
@@ -3302,1438 +2532,2807 @@ impl Parser {
-                crate::TypeInner::Image {
+                ast::Type::Image {
                     dim: crate::ImageDimension::D1,
                     arrayed: true,
                     class: crate::ImageClass::Storage { format, access },
                 }
             }
             "texture_storage_2d" => {
                 let (format, access) = lexer.next_format_generic()?;
-                crate::TypeInner::Image {
+                ast::Type::Image {
                     dim: crate::ImageDimension::D2,
                     arrayed: false,
                     class: crate::ImageClass::Storage { format, access },
                 }
             }
             "texture_storage_2d_array" => {
                 let (format, access) = lexer.next_format_generic()?;
-                crate::TypeInner::Image {
+                ast::Type::Image {
                     dim: crate::ImageDimension::D2,
                     arrayed: true,
                     class: crate::ImageClass::Storage { format, access },
                 }
             }
             "texture_storage_3d" => {
                 let (format, access) = lexer.next_format_generic()?;
-                crate::TypeInner::Image {
+                ast::Type::Image {
                     dim: crate::ImageDimension::D3,
                     arrayed: false,
                     class: crate::ImageClass::Storage { format, access },
                 }
             }
-            _ => return Ok(None),
-        }))
-    }
-
-    const fn check_texture_sample_type(
-        kind: crate::ScalarKind,
-        width: u8,
-        span: Span,
-    ) -> Result<(), Error<'static>> {
-        use crate::ScalarKind::*;
-        // Validate according to https://gpuweb.github.io/gpuweb/wgsl/#sampled-texture-type
-        match (kind, width) {
-            (Float | Sint | Uint, 4) => Ok(()),
-            _ => Err(Error::BadTextureSampleType { span, kind, width }),
+            _ => return Ok(None),
+        }))
+    }
+
+    const fn check_texture_sample_type(
+        kind: crate::ScalarKind,
+        width: u8,
+        span: Span,
+    ) -> Result<(), Error<'static>> {
+        use crate::ScalarKind::*;
+        // Validate according to https://gpuweb.github.io/gpuweb/wgsl/#sampled-texture-type
+        match (kind, width) {
+            (Float | Sint | Uint, 4) => Ok(()),
+            _ => Err(Error::BadTextureSampleType { span, kind, width }),
+        }
+    }
+
+    /// Parse type declaration of a given name.
+    fn parse_type_decl<'a>(
+        &mut self,
+        lexer: &mut Lexer<'a>,
+        mut ctx: ParseExpressionContext<'a, '_, '_>,
+    ) -> Result<Handle<ast::Type<'a>>, Error<'a>> {
+        self.push_rule_span(Rule::TypeDecl, lexer);
+
+        let (name, span) = lexer.next_ident_with_span()?;
+
+        let ty = match self.parse_type_decl_impl(lexer, name, ctx.reborrow())? {
+            Some(ty) => ty,
+            None => ast::Type::User(ast::Ident { name, span }),
```

Comment:
Another missing dependancy:

```suggestion
            None => {
                ctx.unresolved.insert(ast::Dependency {
                    ident: name,
                    usage: span,
                });
                ast::Type::User(ast::Ident { name, span })
            },
```